File: Programmars/Level_3/베스트앨범.py
def solution(genres, plays):
    answer = []
    genre_play = dict()
    for genre, play in zip(genres, plays):
        genre_play[genre] = genre_play.get(genre, 0) + play
    genre_play = sorted(genre_play.items(), key=lambda x: x[1], reverse=True)
    for g in genre_play:
        temp = []
        for i in range(len(genres)):
            if genres[i] == g[0]:
                temp.append((i, plays[i]))
        temp.sort(reverse=True, key=lambda x: x[1])
        if len(temp) == 1:
            answer.append(temp[0][0])
        else:
            for i in range(2):
                answer.append(temp[i][0])
    return answer


print(solution(["classic", "pop", "classic", "classic", "pop"],
               [500, 600, 150, 800, 2500]))  # [4, 1, 3, 0]

File: JavaScript/Programmars/해시/베스트앨범.js
// 문제: https://programmers.co.kr/learn/courses/30/lessons/42579
const solution = (genres, plays) => {
  const genreMap = new Map();
  const countMap = new Map();
  const len = genres.length;
  for (let i = 0; i < len; i++) {
    if (genreMap.has(genres[i])) {
      genreMap.set(genres[i], [...genreMap.get(genres[i]), [i, plays[i]]]);
    } else {
      genreMap.set(genres[i], [[i, plays[i]]]);
    }
    countMap.set(genres[i], (countMap.get(genres[i]) || 0) + plays[i]);
  }

  let countArr = [];
  for (let [genre, count] of countMap) {
    countArr.push([genre, count]);
  }
  countArr.sort((a, b) => b[1] - a[1]);

  const answer = [];
  countArr.forEach(([genre, count]) => {
    const plays = genreMap.get(genre);
    if (plays.length === 1) {
      answer.push(plays[0][0]);
    } else {
      plays.sort((a, b) => {
        if (a[1] === b[1]) return a[0] - b[0];
        return b[1] - a[1];
      });
      answer.push(plays[0][0]);
      answer.push(plays[1][0]);
    }
  });

  return answer;
};

console.log(
  solution(
    ['classic', 'pop', 'classic', 'classic', 'pop'],
    [500, 600, 150, 800, 2500]
  )
); // [4, 1, 3, 0]
